src/dkg/round2.rs

```diff
@@ -231,41 +228,63 @@ where
     P: IntoIterator<Item = &'a round1::PublicPackage>,
     R: RngCore + CryptoRng,
 {
-    let mut round1_public_packages = round1_public_packages.into_iter().collect::<Vec<_>>();
-    round1_public_packages.sort_unstable_by_key(|&p| p.identity());
-    round1_public_packages.dedup();
-    let round1_public_packages = round1_public_packages;
+    let round1_public_packages = round1_public_packages.into_iter().collect::<Vec<_>>();
 
-    let self_public_package = *round1_public_packages
-        .iter()
-        .find(|&p| p.identity() == self_identity)
-        .expect("missing public package for self_identity");
+    // Extract the min/max signers from the secret package
+    let (min_signers, max_signers) = round1::get_secret_package_signers(round1_secret_package);
 
-    let mut frost_packages: BTreeMap<Identifier, Round1Package> = BTreeMap::new();
-    let mut group_secret_key_shards: Vec<&GroupSecretKeyShard> = Vec::new();
+    // Ensure that the number of public packages provided matches max_signers
+    if round1_public_packages.len() != max_signers as usize {
+        return Err(Error::InvalidInput(format!(
+            "expected {} public packages, got {}",
+            max_signers,
+            round1_public_packages.len()
+        )));
+    }
 
-    for public_package in round1_public_packages {
-        if public_package.checksum() != self_public_package.checksum() {
+    // Compute the expected checksum for the public packages
+    let expected_checksum = round1::input_checksum(
+        min_signers,
+        round1_public_packages.iter().map(|pkg| pkg.identity()),
+    );
+
+    // Build the BTree of FROST public packages from our public packages, making sure that the
+    // checksums match, and that every identity was used only once
+    let mut frost_packages = BTreeMap::new();
+    for public_package in &round1_public_packages {
+        if public_package.checksum() != expected_checksum {
             return Err(Error::ChecksumError(ChecksumError::DkgPublicPackageError));
         }
 
-        group_secret_key_shards.push(public_package.group_secret_key_shard());
-
-        // self_public_package must be excluded from frost::keys::dkg::part2 inputs
-        if public_package.identity() == self_identity {
-            continue;
+        if frost_packages
+            .insert(
+                public_package.identity().to_frost_identifier(),
+                public_package.frost_package().clone(),
+            )
+            .is_some()
+        {
```

Comment:
nit - i'm new to rust, but this was confusing for me at first; i had to check that `insert` returns `None` if the key was not already present

Reply:
To be fair, I would have written it as:

```rust
        if let Some(_) = frost_packages
            .insert(
                public_package.identity().to_frost_identifier(),
                public_package.frost_package().clone(),
            )
```

... which IMO makes it much easier to spot that we're checking for something here. But clippy prefers `.is_some()` instead.

Not sure if it would make things easier to understand though.

I could change this to use `.entry()` instead of `.insert()`. The code would become:

```rust
        match frost_packages.entry(public_package.identity().to_frost_identifier()) {
            btree_map::Entry::Vacant(entry) => {
                entry.insert(public_package.frost_package().clone());                
            }   
            btree_map::Entry::Occupied(_) => {                      
                return Err(Error::InvalidInput(format!(
                    "multiple public packages provided for identity {}",
                    public_package.identity()
                )));
            }
        }
```

The performance is the same for every solution listed here. Which one do you like the most?

Reply:
i think the current implementation looks good!

i was just more familiar with something like `if frost_packages.contains_key(...)` 😅